### PR TITLE
Add safe Path.With_check.mkdir_p

### PR DESCRIPTION
**Pull request by Alizter**
We add a `Path.With_check.mkdir_p` which does an extra syscall when the file exists to make sure it is actually a directory.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Could you separate the PR outlining the issues with mkdir_p with unit tests in a separate PR?

**Comment by Alizter** [comment]
@rgrinberg OK, I will include the spec in a later PR.

File: otherlibs/stdune/src/temp.ml [hunk 1]
```diff
@@ -44,6 +44,7 @@ let create_temp_dir ?perms path =
   match Fpath.mkdir ?perms dir with
   | Created -> Ok ()
   | Already_exists -> Error `Retry
+  | Not_a_directory -> Error `Retry
```
Comment (Alizter) [inline otherlibs/stdune/src/temp.ml:47]: I am unsure about this but it seemed like the sensible behaviour to me.

**Comment by Alizter** [comment]
@rgrinberg You mentioned there was a cram test for this exact case already however I could not find it so I kept this cram test.

**Comment by rgrinberg** [comment]
This is actually a useful bug fix. It just needs to be finished

**Comment by Alizter** [comment]
Not gonna finish it soon however so drafting.

**Pull request closed without merging**